docs/src/main/sphinx/develop/tests.md

```diff
@@ -0,0 +1,440 @@
+# Test writing guidelines
+
+The Trino project aims to provide high quality artifacts and features to all
+users. A suite of tests used for development, continuous integration builds,
+and releases is critical for achieving this goal.
+
+Trino acts as an integration layer for many different data sources with the help
+of different connector plugin, and includes support of other plugins as well.
+This results in the need testing many different use cases with a number of
+external systems. The Trino test suite limits the executed tests for any changes
+in pull requests as much as possible, while any merges to the master branch
+process all tests.
+
+The following requirements and guidelines help contributors to create tests with
+the following goals:
+
+* Fast execution.
+* Low requirements in terms of hardware, software, and compute power, and
+  therefore low cost.
+* Reliable execution and test results across multiple invocations.
+* Complex enough to have enough coverage of the execution paths, while avoiding
+  unnecessary complexity.
+* Enable quick, iterative development on developer workstation.
+
+## Requirements
+
+A number of requirements apply to all new tests, and any refactoring work of
+existing tests:
+
+* All tests must use JUnit 5.
+* All tests must use statically imported AssertJ assertions, typically from
+  `org.assertj.core.api.Assertions`.
+* Test class names must start with `Test`, for example `TestExample`
+* Test classes should be defined as package-private and final.
+* Test method must start with `test`, for example `testExplain()`
+* Test methods should be defined as package-private.
+* Tests must be written as unit tests when possible, product or other
+  integration tests should be avoided. These tests typically rely on external
```

Comment:
We have integration tests that use TestContainers, so they are self-contained and don't rely on external infra.